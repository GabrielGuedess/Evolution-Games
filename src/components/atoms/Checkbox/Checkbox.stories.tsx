import { Story, Meta } from '@storybook/react/types-6-0';

import { PlatformIcon } from '../PlatformIcon/PlatformIcon';
import { Checkbox, CheckboxProps } from './Checkbox';

export default {
  title: 'Atoms/Checkbox',
  component: Checkbox,
  args: {
    title: 'Playstation',
    icon: <PlatformIcon platform="playstation" size="small" />,
  },
} as Meta<CheckboxProps>;

export const Default: Story<CheckboxProps> = args => <Checkbox {...args} />;
