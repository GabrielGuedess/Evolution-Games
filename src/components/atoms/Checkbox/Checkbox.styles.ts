import { rgba } from 'polished';

import styled, { css } from 'styled-components';

export const Wrapper = styled.div``;

export const Title = styled.span`
  ${({ theme }) => css`
    color: ${theme.colors.white};
    font-size: ${theme.font.sizes.small};
    transition: color 0.3s;
  `}
`;

export const CheckboxWrapper = styled.label`
  cursor: pointer;
  display: flex;
  align-items: center;
  gap: 2rem;
  padding: 0.375rem;
  position: relative;
  user-select: none;
  margin-bottom: 0.2rem;
`;

export const Icon = styled.svg`
  ${({ theme }) => css`
    width: 2rem;
    height: 2rem;
    background: rgba(113, 123, 138, 0.5);
    border-radius: 0.2rem;

    rect {
      stroke: transparent;
      transition: stroke 0.3s;
    }

    .check-worm,
    .box-worm {
      stroke: ${theme.colors.white};
      filter: drop-shadow(0 0 1rem ${theme.colors.white});
      transition: stroke 0.3s,
        stroke-dashoffset 0.6s cubic-bezier(0.42, -0.2, 0.58, 1.2);
    }
  `}
`;

export const WrapperInfos = styled.div`
  display: flex;
  gap: 1rem;

  ::after {
    border-radius: 0.375rem;
    content: '';
    display: block;
    position: absolute;
    inset: 0;
    z-index: -1;
    transition: background-color 0.15s linear;
  }
`;

export const Checkbox = styled.input`
  ${({ theme }) => css`
    position: absolute;
    top: 0;
    left: 0;
    width: 2rem;
    height: 2rem;
    appearance: none;
    cursor: pointer;
    margin: 0.375rem;

    :checked + ${Icon} {
      background-color: ${theme.colors.primary};
      box-shadow: 0 0.5rem 1.5rem ${rgba(theme.colors.primary, 0.25)};
    }

    :is(:focus, :hover) ~ ${WrapperInfos}:after {
      background-color: ${rgba(theme.colors.primary, 0.15)};
    }

    :is(:checked, :indeterminate) + ${Icon} .box-worm {
      stroke-dashoffset: -91;
    }

    :is(:checked, :indeterminate) + ${Icon} .check-worm {
      stroke-dashoffset: -6;
    }

    :indeterminate + ${Icon} .box-worm {
      stroke-dashoffset: -111.38;
    }

    :indeterminate + ${Icon} .check-worm {
      stroke-dashoffset: -26.38;
    }
  `}
`;
