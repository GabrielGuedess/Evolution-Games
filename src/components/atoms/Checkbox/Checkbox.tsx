import * as S from './Checkbox.styles';

export type CheckboxProps = {
  title: string;
  icon?: JSX.Element;
};

export const Checkbox = ({ title, icon }: CheckboxProps) => (
  <S.Wrapper>
    <S.CheckboxWrapper htmlFor={title}>
      <S.Checkbox type="checkbox" id={title} />

      <S.Icon viewBox="0 0 24 24">
        <g
          fill="none"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
        >
          <rect x="1" y="1" width="22" height="22" />

          <polyline
            className="box-worm"
            points="23,1 1,1 1,23 23,23 23,4"
            strokeDasharray="30 146"
            strokeDashoffset="30"
          />
          <polyline
            className="check-worm"
            points="23,4 10,17 5,12 18,12"
            strokeDasharray="17.38 149.68"
            strokeDashoffset="103.38"
          />
        </g>
      </S.Icon>

      <S.WrapperInfos>
        {icon}

        <S.Title>{title}</S.Title>
      </S.WrapperInfos>
    </S.CheckboxWrapper>
  </S.Wrapper>
);
