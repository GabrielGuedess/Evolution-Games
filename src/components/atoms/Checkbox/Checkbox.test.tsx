import { renderWithTheme } from 'utils/tests/helpers';

import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { Checkbox } from './Checkbox';

describe('<Checkbox />', () => {
  it('should render the heading', () => {
    // Arrange
    const { container } = renderWithTheme(<Checkbox title="test" />);

    // Assert
    expect(screen.getByText('test')).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render dispatch onCheck when status change', async () => {
    const onCheck = jest.fn();

    renderWithTheme(<Checkbox title="checkbox" onCheck={onCheck} />);

    expect(onCheck).not.toHaveBeenCalled();

    userEvent.click(screen.getByRole('checkbox'));
    await waitFor(() => {
      expect(onCheck).toHaveBeenCalledTimes(1);
    });

    expect(onCheck).toBeCalledWith(true);
  });
});
