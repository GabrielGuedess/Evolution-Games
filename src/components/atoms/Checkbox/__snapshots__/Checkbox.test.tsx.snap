// Jest Snapshot v1, https://goo.gl/fbAQLP

exports['<Checkbox /> should render the heading 1'] = `
.c6 {
  color: #FFFFFF;
  font-size: 1.4rem;
  -webkit-transition: color 0.3s;
  transition: color 0.3s;
}

.c0 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 2rem;
  padding: 0.375rem;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  margin-bottom: 0.2rem;
}

.c3 {
  width: 2rem;
  height: 2rem;
  background: rgba(113,123,138,0.5);
  border-radius: 0.2rem;
}

.c3 rect {
  stroke: transparent;
  -webkit-transition: stroke 0.3s;
  transition: stroke 0.3s;
}

.c3 .check-worm,
.c3 .box-worm {
  stroke: #FFFFFF;
  -webkit-filter: drop-shadow(0 0 1rem #FFFFFF);
  filter: drop-shadow(0 0 1rem #FFFFFF);
  -webkit-transition: stroke 0.3s, stroke-dashoffset 0.6s cubic-bezier(0.42,-0.2,0.58,1.2);
  transition: stroke 0.3s, stroke-dashoffset 0.6s cubic-bezier(0.42,-0.2,0.58,1.2);
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 1rem;
}

.c5::after {
  border-radius: 0.375rem;
  content: '';
  display: block;
  position: absolute;
  inset: 0;
  z-index: -1;
  -webkit-transition: background-color 0.15s linear;
  transition: background-color 0.15s linear;
}

.c1 {
  position: absolute;
  top: 0;
  left: 0;
  width: 2rem;
  height: 2rem;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  cursor: pointer;
  margin: 0.375rem;
}

.c1:checked + .c2 {
  background-color: #7C3AED;
  box-shadow: 0 0.5rem 1.5rem rgba(124,58,237,0.25);
}

.c1:is(:focus,:hover) ~ .c4:after {
  background-color: rgba(124,58,237,0.15);
}

.c1:is(:checked,:indeterminate) + .c2 .box-worm {
  stroke-dashoffset: -91;
}

.c1:is(:checked,:indeterminate) + .c2 .check-worm {
  stroke-dashoffset: -6;
}

.c1:indeterminate + .c2 .box-worm {
  stroke-dashoffset: -111.38;
}

.c1:indeterminate + .c2 .check-worm {
  stroke-dashoffset: -26.38;
}

<div
  class=""
>
  <label
    class="c0"
    for="test"
  >
    <input
      class="c1"
      id="test"
      type="checkbox"
    />
    <svg
      class="c2 c3"
      viewBox="0 0 24 24"
    >
      <g
        fill="none"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
      >
        <rect
          height="22"
          width="22"
          x="1"
          y="1"
        />
        <polyline
          class="box-worm"
          points="23,1 1,1 1,23 23,23 23,4"
          stroke-dasharray="30 146"
          stroke-dashoffset="30"
        />
        <polyline
          class="check-worm"
          points="23,4 10,17 5,12 18,12"
          stroke-dasharray="17.38 149.68"
          stroke-dashoffset="103.38"
        />
      </g>
    </svg>
    <div
      class="c4 c5"
    >
      <span
        class="c6"
      >
        test
      </span>
    </div>
  </label>
</div>
`;
